### DISCUSSION TITLE
SMARTS pattern repetition support

### DISCUSSION THREAD

**Discussion by kulmes** [rdkit/rdkit#6000] [Q&A]
I'm attempting to write a SMARTS query that matches an R group of any alkane <C6 in a particular position of a molecule.

I'm quite familiar with regular expressions but not as much with the SMARTS syntax.  I'm attempting to do something like this :
patt = '[CX4]{1-5}' (I think my problem here is that I can't specify that these are the only atoms that can be matched ... eg. match anything with only 1-5 carbons that have 4 bonds inclusive of H bonds)

or '[$([CX4]{1-5})&!$([!C];[CX{1-3})]' (I'm still trying to figure out how to apply SMARTS logic operators, so this one is likely messed up

And have it match:
C
CC
CCCCC
C(C)C(CC)
C1CCCC1
CC1CCC1
etc...

But not
CCCCCC 
C=C
C#C
CCCC=C
N
C-N
S
P
etc...

I'm assuming I'm doing something wrong with the curly bracket number of pattern specification, I can't find a lot of information on pattern repetition specification syntax for different SMARTS implementations.

**Comment by greglandrum** [comment]
@kulmes : this isn't something you can do with regular SMARTS: you cannot have a single SMARTS pattern which matches variable numbers of atoms. 

The RDKit does support SMARTS extensions (CXSMARTS) which allow you to specify link nodes: atoms which can be repeated a variable number of times in the query. These look like this:
`O-[CX4]-[CH3] |LN:1:1.5|`
So that's an O connected via a single bond to 1-5 `[CX4]` atoms connected via a single bond to a `[CH3]`.

Here's an example of how that works in an example like yours:
```
In [1]: from rdkit import Chem
In [2]: qry = Chem.MolFromSmarts('O-[CX4]-[CH3] |LN:1:1.5|')
In [3]: from rdkit.Chem import rdMolEnumerator
In [4]: bndl = rdMolEnumerator.Enumerate(qry)
In [5]: Chem.MolFromSmiles('OC').HasSubstructMatch(bndl)
Out[5]: False
In [6]: Chem.MolFromSmiles('OCC').HasSubstructMatch(bndl)
Out[6]: True
In [7]: Chem.MolFromSmiles('OCCCCCC').HasSubstructMatch(bndl)
Out[7]: True
In [8]: Chem.MolFromSmiles('OCCCCCCC').HasSubstructMatch(bndl)
Out[8]: False
In [9]: Chem.MolFromSmiles('OC(C)CC').HasSubstructMatch(bndl)
Out[9]: True
```
*NOTE:* due to a bug in the current RDKit release (#6014), this will not work for you until the 2022.09.5 release comes in a few weeks.

There's more information on link nodes in this blog post; https://greglandrum.github.io/rdkit-blog/posts/2021-08-03-generalized-substructure-search.html

**Reply by kulmes** [reply]
Thanks!  That seems like it will get me all branched alkanes.  I'm trying to work my way through your patch to see if it will also get me any cyclic alkanes ... e.g. would these also evaluate to true
Chem.MolFromSmiles('O-C1CCCC1').HasSubstructMatch(bndl)
Chem.MolFromSmiles('O-CC1CCC1').HasSubstructMatch(bndl)

I'm thinking not since they don't have a terminal [CH3].  I'm also finding that at least in the current implementation that I can't specify a pattern repetition of 0, so 'OC' does not match qry = Chem.MolFromSmarts('O-[CX4]-[CH3] |LN:1:1.5|') and I can't specify either :
qry = Chem.MolFromSmarts('O-[CX4]-[CH3] |LN:1:0.5|')
qry = Chem.MolFromSmarts('O-[CX4] |LN:1:1.5|')

That is likely more my not being as familiar with CXSMARTS.  I'm reading what I can find on the specification, but a lot of the good documentation and examples seem to be either your blog posts or more related to CXSMILES (which is similar enough)

I'm also running into something odd:
patt = 'C(=[OX1])(-O-[CX4]-[CH3])([$([CH3]),$(C[CH3]),$(CC[CH3]),$(C([CH3])[CH3])]) |LN:3:1.5|'
bndl = rdMolEnumerator.Enumerate(Chem.MolFromSmarts(patt))
bndl
is doing the enumeration for the link node but not for the recursive SMARTS options ... so:
Chem.MolFromSmiles('CCOC(=O)(C(C)C)').HasSubstructMatch(bndl) = FALSE
Chem.MolFromSmiles('CCOC(=O)(C(C)C)').HasSubstructMatch(Chem.MolFromSmarts(patt) = TRUE (but this is only TRUE for the 2 carbon chain because it's not doing the link node expansion as match options)

I went down this route because I couldn't figure out how to get the link node syntax to match just a single methyl or an alkyl chain of variable length, so was going to specify either a single methyl or the '[CX4]-[CH3] ..." w/ link node expansion that you provided above, but I can't seem to get either expansion of all possible options into a bundle or pattern matching to work.  I'm sure I can solve this problem with a much more complex loop enumerating all the options, but I'm trying to keep the specification of the pattern as simple as possible
and I figured there would be some easy way of saying there are up to 6 carbon atoms hanging off this attachment point all connected with only single bonds in any configuration.

Anyways I appreciate the help.  I'm going to go digging for some more in depth documentation of CXSMARTS and/or examples.
